### ISSUE TITLE
Child - Controller getActivity() is null because hostcontroller is null

### ISSUE THREAD

**Issue by sockeqwe** [bluelinelabs/Conductor#85]
Hi, 
I struggled upon the following problem:

I have added a `LifecycleListener` (not sure if that matters) and I listen for `preDestroyView()` events.

In `preDestroyView(Controller, View)` of a "regular" (not child controller)
 Controller everything works as expected. `controller.getActivity()` returns the activity which is not null (controller is first argument of `preDestroyView()`).

If I attach the same LifecycleListener to a "child" controller (on a child router), then `controller.getActivity()` will return null, because `router.getActivity()` returns null, because `ControllerHosterRouter.getActivity()` returns null because `ControllerHosterRouter.hostcontroller` is null.

It seems like `ControllerHosterRouter.removeHost()` will be called in `Controller.detach()`.
I was wondering whether or not this is the expected behavior? If yes, shouldn't this behavior for a "regular" (not child controller) be the same?

The problem seems to be, that `Controller.removeViewRef()` will loop over the child controllers and set `ControllerHosterRouter.removeHost()`, but the child controller's `onDestroyView()` (and corresponding LifecycleListeners) will be called in next  loop recursively.

https://github.com/bluelinelabs/Conductor/blob/develop/conductor/src/main/java/com/bluelinelabs/conductor/Controller.java#L786

In general I guess the implementation corresponds to the docs `Controller.onDettach()` (although it has non consistent behavior between "regular" and "child" controller at this point):

> Called when this Controller is detached from its host ViewGroup

So does this also mean that `onAttach()` and `onDetach()` means attached and detached from Activity too? Nevertheless, if I understood your code correctly we run into the same problem: `ControllerHosterRouter.removeHost()` will be called for the child controllers, before the actual child controller will be detached (corresponding callbacks / LifecycleListener callbacks invoked).

**Comment by EricKuck** [comment]
I haven't had a chance to dig into every possible scenario where this could potentially happen, but the latest snapshot contains a fix that I *think* will cover everything. Let me know if anything wasn't covered!

**Comment by chris6647** [comment]
Thanks for being so fast @EricKuck ! I wanted to test if it fixed the issue in my app, but gradle doesnt seem to be able to detect your latest SNAPSHOT:
`Error:Could not find com.bluelinelabs:conductor:2.0.1-SNAPSHOT.`
I just updated the dependency to:
` compile 'com.bluelinelabs:conductor:2.0.1-SNAPSHOT`
as your README says, but to no avail.

**Comment by sockeqwe** [comment]
@chris6647 you have to add the url to the snapshot repository:
```groovy
maven { url "https://oss.sonatype.org/content/repositories/snapshots/" } 
```

For instance in your "main" build.gradle file:

```groovy
allprojects {
  repositories {
    jcenter()

    maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
}
```

**Comment by chris6647** [comment]
@sockeqwe I had a feeling i was noobing around when i wrote that previous post. Tried some other urls, but didnt know that specific one, so thanks man :)

**Comment by sockeqwe** [comment]
Seems to be fixed in `2.0.1-SNAPSHOT`. Thanks!
